src/hotspot/share/gc/g1/g1HeapSizingPolicy.cpp

```diff
@@ -436,3 +448,150 @@ size_t G1HeapSizingPolicy::full_collection_resize_amount(bool& expand, size_t al
+        _candidates->append(r);
+        (*_inactive_regions)++;
+      }
+      return false;
+    }
+  } cl(candidates, &inactive_regions, this);
+
+  _g1h->heap_region_iterate(&cl);
+
+  if (inactive_regions > 0) {
+    log_debug(gc, sizing)("Uncommit candidates found: %u inactive regions out of %u total regions",
+                  inactive_regions, _g1h->max_num_regions());
+    log_debug(gc, sizing)("Region state transition: %u regions found eligible for uncommit after scan",
+                  inactive_regions);
+  }
+  log_debug(gc, sizing)("Full region scan: found %u inactive regions out of %u total regions",
+                       inactive_regions,
+                       _g1h->max_num_regions());
+}
+
+bool G1HeapSizingPolicy::should_uncommit_region(G1HeapRegion* hr) const {
+  // Note: Caller already guarantees hr->is_empty() is true
+  // Empty regions should always be free and not in collection set in normal operation
+
+  jlong current_time = os::javaTimeMillis();
+  jlong last_access = hr->last_access_time();
+  jlong elapsed = current_time - last_access;
+
+  log_trace(gc, sizing)("Region %u uncommit check: elapsed=" JLONG_FORMAT "ms threshold=" JLONG_FORMAT "ms last_access=" JLONG_FORMAT " now=" JLONG_FORMAT " empty=%s",
+                     hr->hrm_index(), elapsed, (jlong)_uncommit_delay_ms, last_access, current_time,
+                     hr->is_empty() ? "true" : "false");
+
+  bool should_uncommit = elapsed > (jlong)_uncommit_delay_ms;
+  if (should_uncommit) {
+    log_debug(gc, sizing)("Region state transition: Region %u transitioning from active to inactive after " JLONG_FORMAT "ms idle",
+                  hr->hrm_index(), elapsed);
+  }
+
+  return should_uncommit;
+}
+
+size_t G1HeapSizingPolicy::evaluate_heap_resize(bool& expand) {
+  expand = false; // Time-based sizing only handles uncommit, never expansion
+
+  if (!G1UseTimeBasedHeapSizing) {
+    return 0;
+  }
+
+  // Don't resize during GC
+  if (_g1h->is_stw_gc_active()) {
+    return 0;
+  }
+
+  // Must hold Heap_lock during heap resizing
+  MutexLocker ml(Heap_lock);
+
+  ResourceMark rm; // Ensure GrowableArray resources are properly released
+
+  // Find regions eligible for uncommit
+  GrowableArray<G1HeapRegion*> candidates;
+  get_uncommit_candidates(&candidates);
+
+  uint inactive_count = candidates.length();
+  uint total_regions = _g1h->max_num_regions();
+
+  // Need minimum number of inactive regions to proceed
+  if (inactive_count >= G1MinRegionsToUncommit) {
+    size_t region_size = G1HeapRegion::GrainBytes;
+    size_t current_heap = _g1h->capacity();
+    size_t min_heap = MAX2((size_t)InitialHeapSize, MinHeapSize);  // Never go below initial size
+
+    // Calculate maximum bytes we can uncommit while respecting min heap size
+    size_t max_shrink_bytes = current_heap > min_heap ? current_heap - min_heap : 0;
+
+    log_trace(gc, sizing)("Time-based evaluation details: current_heap=%zuB min_heap=%zuB "
```

Comment:
I do not think `trace`log messages need the "details" in `Time-based evaluation details" - this is already indicated by the log level.

Also, in contrast to other "Time-based evaluation" messages this has a ":" at the end, which is inconsistent.

Reply:
ok changed to:
log_trace(gc, sizing)("Uncommit evaluation: current_heap=%zuB min_heap=%zuB ...")